pacman: sometimes hangs validating GPG signatures

I first saw this on Windows on ARM, but it seems to have shown up on msys2-autobuild recently too (https://github.com/msys2/msys2-autobuild/runs/4552699834?check_suite_focus=true and https://github.com/msys2/msys2-autobuild/runs/4575182521?check_suite_focus=true).

What I saw on ARM was that `pstree` showed two pacman processes, one a child of the other.  If I killed the child, via its winpid and taskkill, pacman would complain about a GPGME error and things would move on (with a failed invocation of pacman).  If I killed the parent, the child would still hang around as an apparent zombie.

I tried to debug it once a while back, but I seem to recall that attempting to attach a debugger to it caused it to 'unblock' and exit, which made trying to figure out what was going on pretty difficult.

In case the job logs expire, they look something like:
```
2021-12-19T16:14:37.5863759Z checking keyring...
2021-12-19T16:14:37.7745146Z checking package integrity...
2021-12-19T21:28:27.2917878Z ##[error]The operation was canceled.
```

--- a/.ci/ci-build.sh
+++ b/.ci/ci-build.sh
@@ -50,7 +50,7 @@ for package in "${packages[@]}"; do
         for pkg in *.pkg.tar.*; do
             pkgname="$(echo "$pkg" | rev | cut -d- -f4- | rev)"
             echo "::group::[install] ${pkgname}"
-            grep -qFx "${package}" "$(dirname "$0")/ci-dont-install-list.txt" || pacman --noprogressbar --upgrade --noconfirm $pkg
+            grep -qFx "${package}" "$DIR/ci-dont-install-list.txt" || pacman --noprogressbar --upgrade --noconfirm $pkg
             echo "::endgroup::"
 
             echo "::group::[diff] ${pkgname}"
@@ -71,7 +71,7 @@ for package in "${packages[@]}"; do
         cd - > /dev/null
     else
         echo "::group::[install] ${package}"
-        grep -qFx "${package}" "$(dirname "$0")/ci-dont-install-list.txt" || execute 'Installing' install_packages
+        grep -qFx "${package}" "$DIR/ci-dont-install-list.txt" || execute 'Installing' install_packages
         echo "::endgroup::"
 
         echo "::group::[diff] ${package}"
@@ -101,7 +101,7 @@ for package in "${packages[@]}"; do
         for pkg in *.pkg.tar.*; do
             installed_packages+=("$(echo "$pkg" | rev | cut -d- -f4- | rev)")
         done
-        pacman -R --recursive --unneeded --noconfirm --noprogressbar "${installed_packages[@]}"
+        grep -qFx "${package}" "$DIR/ci-dont-install-list.txt" || pacman -R --recursive --unneeded --noconfirm --noprogressbar "${installed_packages[@]}"
         unset installed_packages
         cd - > /dev/null
         echo "::endgroup::"

--- a/msys2-runtime/0036-Cygwin-path-Convert-UNC-path-prefix-back-to-drive-le.patch
+++ b/msys2-runtime/0036-Cygwin-path-Convert-UNC-path-prefix-back-to-drive-le.patch
@@ -1,0 +1,70 @@
+From 0efcde460df9c5d7e7e4bc52bc134a817a8c33aa Mon Sep 17 00:00:00 2001
+From: Takashi Yano <takashi.yano@nifty.ne.jp>
+Date: Wed, 8 Dec 2021 20:15:45 +0900
+Subject: [PATCH 36/N] Cygwin: path: Convert UNC path prefix back to drive
+ letter.
+
+- When symlink_info::check() is called with the path having drive
+  letter and UNC path is mounted to the drive, the path is replaced
+  with UNC path. With this patch, UNC path prefix is converted back
+  to drive letter.  This fixes the issue:
+  https://cygwin.com/pipermail/cygwin/2021-November/250087.html
+  https://cygwin.com/pipermail/cygwin/2021-December/250103.html
+---
+ winsup/cygwin/path.cc | 37 ++++++++++++++++++++++++++++++++++++-
+ 1 file changed, 36 insertions(+), 1 deletion(-)
+
+diff --git a/winsup/cygwin/path.cc b/winsup/cygwin/path.cc
+index 3f24217..06b26d2 100644
+--- a/winsup/cygwin/path.cc
++++ b/winsup/cygwin/path.cc
+@@ -3655,10 +3655,45 @@ restart:
+ 	    {
+ 	      UNICODE_STRING fpath;
+ 
+-	      RtlInitCountedUnicodeString (&fpath, fpbuf, ret * sizeof (WCHAR));
++	      /* If incoming path has no trailing backslash, but final path
++		 has one, drop trailing backslash from final path so the
++		 below string comparison has a chance to succeed. */
++	      if (upath.Buffer[(upath.Length - 1) / sizeof (WCHAR)] != L'\\'
++                  && fpbuf[ret - 1] == L'\\')
++                fpbuf[--ret] = L'\0';
+ 	      fpbuf[1] = L'?';	/* \\?\ --> \??\ */
++	      RtlInitCountedUnicodeString (&fpath, fpbuf, ret * sizeof (WCHAR));
+ 	      if (!RtlEqualUnicodeString (&upath, &fpath, !!ci_flag))
+ 	        {
++		  /* Check if the final path is an UNC path and the incoming
++		     path isn't.  If so... */
++		  if (RtlEqualUnicodePathPrefix (&fpath, &ro_u_uncp, TRUE)
++		      && !RtlEqualUnicodePathPrefix (&upath, &ro_u_uncp, TRUE))
++		    {
++		      /* ...get the remote path from the volume path name,
++			 replace remote path with drive letter, check again. */
++		      WCHAR remote[MAX_PATH];
++
++		      fpbuf[1] = L'\\';
++		      BOOL r = GetVolumePathNameW (fpbuf, remote, MAX_PATH);
++		      fpbuf[1] = L'?';
++		      if (r)
++			{
++			  int remlen = wcslen (remote);
++			  if (remote[remlen - 1] == L'\\')
++			    remlen--;
++			  /* Hackfest */
++			  fpath.Buffer[4] = upath.Buffer[4]; /* Drive letter */
++			  fpath.Buffer[5] = L':';
++			  WCHAR *to = fpath.Buffer + 6;
++			  WCHAR *from = to + remlen - 6;
++			  memmove (to, from,
++				   (wcslen (from) + 1) * sizeof (WCHAR));
++			  fpath.Length -= (from - to) * sizeof (WCHAR);
++			  if (RtlEqualUnicodeString (&upath, &fpath, !!ci_flag))
++			    goto file_not_symlink;
++			}
++		    }
+ 		  issymlink = true;
+ 		  /* upath.Buffer is big enough and unused from this point on.
+ 		     Reuse it here, avoiding yet another buffer allocation. */
+-- 
+2.34.1.windows.1
+

--- a/msys2-runtime/0037-Cygwin-path-Fix-path-conversion-of-virtual-drive.patch
+++ b/msys2-runtime/0037-Cygwin-path-Fix-path-conversion-of-virtual-drive.patch
@@ -1,0 +1,85 @@
+From 2480c0e3242665890e1dac01891c5a7834fdd138 Mon Sep 17 00:00:00 2001
+From: Takashi Yano <takashi.yano@nifty.ne.jp>
+Date: Thu, 9 Dec 2021 17:15:23 +0900
+Subject: [PATCH 37/N] Cygwin: path: Fix path conversion of virtual drive.
+
+- The last change in path.cc introduced a bug that causes an error
+  when accessing a virtual drive which mounts UNC path such as
+  "\\server\share\dir" rather than "\\server\share". This patch
+  fixes the issue.
+---
+ winsup/cygwin/path.cc | 56 +++++++++++++++++++++++++++----------------
+ 1 file changed, 35 insertions(+), 21 deletions(-)
+
+diff --git a/winsup/cygwin/path.cc b/winsup/cygwin/path.cc
+index 06b26d2..36f15bd 100644
+--- a/winsup/cygwin/path.cc
++++ b/winsup/cygwin/path.cc
+@@ -3670,29 +3670,43 @@ restart:
+ 		  if (RtlEqualUnicodePathPrefix (&fpath, &ro_u_uncp, TRUE)
+ 		      && !RtlEqualUnicodePathPrefix (&upath, &ro_u_uncp, TRUE))
+ 		    {
+-		      /* ...get the remote path from the volume path name,
+-			 replace remote path with drive letter, check again. */
++		      /* ...get the remote path, replace remote path
++			 with drive letter, check again. */
++		      WCHAR drive[3] =
++			{(WCHAR) towupper (upath.Buffer[4]), L':', L'\0'};
+ 		      WCHAR remote[MAX_PATH];
+ 
+-		      fpbuf[1] = L'\\';
+-		      BOOL r = GetVolumePathNameW (fpbuf, remote, MAX_PATH);
+-		      fpbuf[1] = L'?';
+-		      if (r)
+-			{
+-			  int remlen = wcslen (remote);
+-			  if (remote[remlen - 1] == L'\\')
+-			    remlen--;
+-			  /* Hackfest */
+-			  fpath.Buffer[4] = upath.Buffer[4]; /* Drive letter */
+-			  fpath.Buffer[5] = L':';
+-			  WCHAR *to = fpath.Buffer + 6;
+-			  WCHAR *from = to + remlen - 6;
+-			  memmove (to, from,
+-				   (wcslen (from) + 1) * sizeof (WCHAR));
+-			  fpath.Length -= (from - to) * sizeof (WCHAR);
+-			  if (RtlEqualUnicodeString (&upath, &fpath, !!ci_flag))
+-			    goto file_not_symlink;
+-			}
++
++		      int remlen = QueryDosDeviceW (drive, remote, MAX_PATH);
++		      if (remlen < 3)
++			goto file_not_symlink; /* fallback */
++		      remlen -= 2;
++
++		      if (remote[remlen - 1] == L'\\')
++			remlen--;
++		      WCHAR *p;
++		      UNICODE_STRING rpath;
++		      RtlInitCountedUnicodeString (&rpath, remote,
++						   remlen * sizeof (WCHAR));
++		      if (RtlEqualUnicodePathPrefix (&rpath, &ro_u_uncp, TRUE))
++			remlen -= 6;
++		      else if ((p = wcschr (remote, L';'))
++			       && p + 3 < remote + remlen
++			       && wcsncmp (p + 1, drive, 2) == 0
++			       && (p = wcschr (p + 3, L'\\')))
++			remlen -= p - remote - 1;
++		      else
++			goto file_not_symlink; /* fallback */
++		      /* Hackfest */
++		      fpath.Buffer[4] = drive[0]; /* Drive letter */
++		      fpath.Buffer[5] = L':';
++		      WCHAR *to = fpath.Buffer + 6;
++		      WCHAR *from = to + remlen;
++		      memmove (to, from,
++			       (wcslen (from) + 1) * sizeof (WCHAR));
++		      fpath.Length -= (from - to) * sizeof (WCHAR);
++		      if (RtlEqualUnicodeString (&upath, &fpath, !!ci_flag))
++			goto file_not_symlink;
+ 		    }
+ 		  issymlink = true;
+ 		  /* upath.Buffer is big enough and unused from this point on.
+-- 
+2.34.1.windows.1
+

--- a/msys2-runtime/0038-fixup-CI-add-a-GHA-for-doing-a-basic-build-test.patch
+++ b/msys2-runtime/0038-fixup-CI-add-a-GHA-for-doing-a-basic-build-test.patch
@@ -1,0 +1,26 @@
+From f2fd51bcb843ff44060b941e5f1b58f97b134a14 Mon Sep 17 00:00:00 2001
+From: jeremyd2019 <github@jdrake.com>
+Date: Wed, 22 Dec 2021 19:17:32 -0800
+Subject: [PATCH 38/N] fixup! CI: add a GHA for doing a basic build test
+
+CI now needs to depend on autotools
+---
+ .github/workflows/build.yaml | 2 +-
+ 1 file changed, 1 insertion(+), 1 deletion(-)
+
+diff --git a/.github/workflows/build.yaml b/.github/workflows/build.yaml
+index 8b0817c..8c8fd3e 100644
+--- a/.github/workflows/build.yaml
++++ b/.github/workflows/build.yaml
+@@ -15,7 +15,7 @@ jobs:
+         with:
+           msystem: MSYS
+           update: true
+-          install: msys2-devel base-devel cocom diffutils gcc gettext-devel libiconv-devel make mingw-w64-cross-crt mingw-w64-cross-gcc mingw-w64-cross-zlib perl zlib-devel
++          install: msys2-devel base-devel autotools cocom diffutils gcc gettext-devel libiconv-devel make mingw-w64-cross-crt mingw-w64-cross-gcc mingw-w64-cross-zlib perl zlib-devel
+ 
+       - name: Build
+         shell: msys2 {0}
+-- 
+2.34.1.windows.1
+

--- a/msys2-runtime/0039-Cygwin-fhandler_pipe-get_query_hdl_per_process-avoid.patch
+++ b/msys2-runtime/0039-Cygwin-fhandler_pipe-get_query_hdl_per_process-avoid.patch
@@ -1,0 +1,65 @@
+From e4fbec279b49055a139a8f26ab4601ea30064e2d Mon Sep 17 00:00:00 2001
+From: Ken Brown <kbrown@cornell.edu>
+Date: Sun, 26 Dec 2021 16:42:26 -0500
+Subject: [PATCH 39/N] Cygwin: fhandler_pipe::get_query_hdl_per_process: avoid
+ a crash
+
+NtQueryInformationProcess(ProcessHandleInformation) can return
+STATUS_SUCCESS with invalid handle data for certain processes
+("minimal" processes on Windows 10).  This can cause a crash when
+there's an attempt to access that data.  Fix that by setting
+NumberOfHandles to zero before calling NtQueryInformationProcess.
+
+Addresses: https://cygwin.com/pipermail/cygwin-patches/2021q4/011611.html
+---
+ winsup/cygwin/fhandler_pipe.cc | 13 +++++++++++--
+ 1 file changed, 11 insertions(+), 2 deletions(-)
+
+diff --git a/winsup/cygwin/fhandler_pipe.cc b/winsup/cygwin/fhandler_pipe.cc
+index 9ce1400..e4b71b5 100644
+--- a/winsup/cygwin/fhandler_pipe.cc
++++ b/winsup/cygwin/fhandler_pipe.cc
+@@ -6,8 +6,6 @@ This software is a copyrighted work licensed under the terms of the
+ Cygwin license.  Please consult the file "CYGWIN_LICENSE" for
+ details. */
+ 
+-/* FIXME: Should this really be fhandler_pipe.cc? */
+-
+ #include "winsup.h"
+ #include <stdlib.h>
+ #include <sys/socket.h>
+@@ -20,6 +18,7 @@ details. */
+ #include "pinfo.h"
+ #include "shared_info.h"
+ #include "tls_pbuf.h"
++#include <assert.h>
+ 
+ /* This is only to be used for writing.  When reading,
+ STATUS_PIPE_EMPTY simply means there's no data to be read. */
+@@ -1260,6 +1259,12 @@ fhandler_pipe::get_query_hdl_per_process (WCHAR *name,
+ 	    HeapAlloc (GetProcessHeap (), 0, nbytes);
+ 	  if (!phi)
+ 	    goto close_proc;
++	  /* NtQueryInformationProcess can return STATUS_SUCCESS with
++	     invalid handle data for certain processes.  See
++	     https://github.com/processhacker/processhacker/blob/05f5e9fa477dcaa1709d9518170d18e1b3b8330d/phlib/native.c#L5754.
++	     We need to ensure that NumberOfHandles is zero in this
++	     case to avoid a crash in the for loop below. */
++	  phi->NumberOfHandles = 0;
+ 	  status = NtQueryInformationProcess (proc, ProcessHandleInformation,
+ 					      phi, nbytes, &len);
+ 	  if (NT_SUCCESS (status))
+@@ -1271,6 +1276,10 @@ fhandler_pipe::get_query_hdl_per_process (WCHAR *name,
+       if (!NT_SUCCESS (status))
+ 	goto close_proc;
+ 
++      /* Sanity check in case Microsoft changes
++	 NtQueryInformationProcess and the initialization of
++	 NumberOfHandles above is no longer sufficient. */
++      assert (phi->NumberOfHandles <= n_handle);
+       for (ULONG j = 0; j < phi->NumberOfHandles; j++)
+ 	{
+ 	  /* Check for the peculiarity of cygwin read pipe */
+-- 
+2.34.1.windows.1
+

--- a/msys2-runtime/PKGBUILD
+++ b/msys2-runtime/PKGBUILD
@@ -4,7 +4,7 @@
 pkgbase=msys2-runtime
 pkgname=('msys2-runtime' 'msys2-runtime-devel')
 pkgver=3.3.3
-pkgrel=3
+pkgrel=4
 pkgdesc="Cygwin POSIX emulation engine"
 arch=('i686' 'x86_64')
 url="https://www.cygwin.com/"
@@ -57,7 +57,11 @@ source=('msys2-runtime'::git://sourceware.org/git/newlib-cygwin.git#tag=cygwin-$
         0032-docs-skip-building-texinfo-and-PDF-files.patch
         0033-install-libs-depend-on-the-toollibs.patch
         0034-fixup-Rename-dll-from-cygwin-to-msys.patch
-        0035-Cygwin-pty-fix-bug-in-reduced-input-transfer.patch)
+        0035-Cygwin-pty-fix-bug-in-reduced-input-transfer.patch
+        0036-Cygwin-path-Convert-UNC-path-prefix-back-to-drive-le.patch
+        0037-Cygwin-path-Fix-path-conversion-of-virtual-drive.patch
+        0038-fixup-CI-add-a-GHA-for-doing-a-basic-build-test.patch
+        0039-Cygwin-fhandler_pipe-get_query_hdl_per_process-avoid.patch)
 sha256sums=('SKIP'
             'c933090fc69948f47dc6452fb9913750de9e341fd80534e7099c7a16f8391167'
             'c25197a13a4df28bd34283d830d14d1c091b6517a5f298ce0f50b14fd31b7e8f'
@@ -93,7 +97,11 @@ sha256sums=('SKIP'
             '4c3fcaa7f63581488504d321bca7a475153b5a1327e913ddbeb2f80443f16757'
             'db27ac00a4f1f5621cfcbbb215cd2b8af629d347b7d1de56145f18360fb90a58'
             '8257cf1beeb00b95eb1ac0a913de3aa1844ee7ea1b16cee59ecd205cf7deab9f'
-            '8545ac529b2fa7d33b92735c3371a16b0377b5860c94254b78882da70af143f0')
+            '8545ac529b2fa7d33b92735c3371a16b0377b5860c94254b78882da70af143f0'
+            '83d6900c52766855325a37a9f951b856c2015b67c6454f8a6ca0a0faac5f08e1'
+            '115e24f1ffb6986afd594f3e5f074a63291ff9dc53064e30617b97816e838946'
+            '22ad7003f66e3dca60871d4ff9c35d2461d9af72a91c33b622d105b587c0b64d'
+            'e37cb6ef13f15fd9eb21ecf9dec985a5bd0eedb8cfbf656f9932f3e609228982')
 
 # Helper macros to help make tasks easier #
 apply_patch_with_msg() {
@@ -165,7 +173,11 @@ prepare() {
   0032-docs-skip-building-texinfo-and-PDF-files.patch \
   0033-install-libs-depend-on-the-toollibs.patch \
   0034-fixup-Rename-dll-from-cygwin-to-msys.patch \
-  0035-Cygwin-pty-fix-bug-in-reduced-input-transfer.patch
+  0035-Cygwin-pty-fix-bug-in-reduced-input-transfer.patch \
+  0036-Cygwin-path-Convert-UNC-path-prefix-back-to-drive-le.patch \
+  0037-Cygwin-path-Fix-path-conversion-of-virtual-drive.patch \
+  0038-fixup-CI-add-a-GHA-for-doing-a-basic-build-test.patch \
+  0039-Cygwin-fhandler_pipe-get_query_hdl_per_process-avoid.patch
 }
 
 build() {

